### DISCUSSION TITLE
How can I use multiple nodes on HPC?

### DISCUSSION THREAD

**Discussion by mattocci27** [ropensci/targets#1067] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I am trying to run the example in `_targets.R` within a Singularity container on an HPC system (Slurm scheduler). I aim to use multiple nodes since I will eventually need many CPUs.

While `tar_make()` works fine, `tar_make_clustermq()` fails when I set `clustermq.scheduler = "slurm"`. Although `tar_make_clustermq()` works with `clustermq.scheduler = "multisession"`, I want to use multiple nodes.

Do you have any suggestions? Thanks.

```
#_targets.R

library(targets)
library(tarchetypes)
library(clustermq)
library(tibble)

tar_option_set(packages = c("clustermq", "tibble"))

options(
  clustermq.scheduler = "slurm",
  clustermq.template = "slurm_clustermq.tmpl")

long_computation <- function(x) {
  Sys.sleep(5)
  x^2
}

targets <- tar_map(
  values = tibble(id = 1:5),
  tar_target(
    test,
    long_computation(id),
    resources = tar_resources(
      clustermq = tar_resources_clustermq(
        template = list(cpus = 1, walltime = "00:10:00", memory = "1G", partition = "my-partition", nodes = 5)))
    ,
    deployment = "worker"
  )
)

list(targets)
```

```
#slurm_clustermq.tmpl
#!/bin/bash
#SBATCH --job-name={{ job_name }}
#SBATCH --output={{ log_file }}
#SBATCH --error={{ log_file }}
#SBATCH --ntasks=1
#SBATCH --cpus-per-task={{ cpus }}
#SBATCH --time={{ walltime }}
#SBATCH --mem={{ memory }}
#SBATCH --partition={{ partition }}
#SBATCH --nodes=1
#SBATCH --export=NONE
#SBATCH --output=test.out
#SBATCH --error=test.err

module purge
module load singularity/3.7.3

singularity exec --env RENV_PATHS_CACHE=/work/${USER}/renv \
  --env RENV_PATHS_PREFIX_AUTO=TRUE \
  radian.sif R --no-save --no-restore -e 'clustermq:::worker("{{ master }}")' > test.log
```

**Comment by wlandau** [comment]
That Singularity container could be causing the same issue @brendanf describes at https://github.com/wlandau/crew.cluster/issues/1#issuecomment-1539581789. `clustermq` launches workers as jobs which then try to connect back to the controller on the local network. The local IP address that `clustermq` is aware of may be hidden behind that container. Unfortunately I have little experience with containers can can't really recommend a solution at this point, other than the `crew.cluster`-related test I proposed at https://github.com/wlandau/crew.cluster/issues/1#issuecomment-1540135170.
